sklearn/neighbors/classification.py

```diff
@@ -364,27 +381,49 @@ def predict(self, X):
 
         weights = _get_weights(neigh_dist, self.weights)
 
-        y_pred = np.empty((n_samples, n_outputs), dtype=classes_[0].dtype)
-        for k, classes_k in enumerate(classes_):
-            pred_labels = np.array([_y[ind, k] for ind in neigh_ind],
-                                   dtype=object)
-            if weights is None:
-                mode = np.array([stats.mode(pl)[0]
-                                 for pl in pred_labels[inliers]], dtype=np.int)
-            else:
-                mode = np.array([weighted_mode(pl, w)[0]
-                                 for (pl, w)
-                                 in zip(pred_labels[inliers], weights[inliers])],
-                                dtype=np.int)
+        if self.outputs_2d_ == 'sparse':
+            y_pred_sparse_multilabel = []
 
-            mode = mode.ravel()
+            for k, classes_k in enumerate(classes_):
+                pred_labels = np.array([_y[ind, k].toarray()
+                                        for ind in neigh_ind[inliers]],
+                                       dtype=object)
+                y_pred_k = np.zeros(n_samples)
+                mode = self._mode(pred_labels, weights, inliers)
+                y_pred_k[inliers] = classes_k.take(mode)
 
-            y_pred[inliers, k] = classes_k.take(mode)
+                if outliers:
+                    y_pred_k[outliers] = self.outlier_label
 
-        if outliers:
-            y_pred[outliers, :] = self.outlier_label
+                y_pred_sparse_multilabel.append(csc_matrix(y_pred_k).T)
```

Comment:
this doesn't make sense in terms of efficient data structures. I think you mean `csr_matrix(y_pred_k).T`, or you mean `csc_matrix(y_pred_k.reshape(-1, 1))`